src/riemannian_diff.jl

```diff
@@ -0,0 +1,232 @@
+
+"""
+    AbstractRiemannianDiffBackend
+
+An abstract type for diff backends. See [`RiemannianONBDiffBackend`](@ref) for
+an example.
+"""
+abstract type AbstractRiemannianDiffBackend end
+
+"""
+    differential(f::Any, c::Curve, t::Real, backend::AbstractDiffBackend = rdifferential_backend())
+
+Compute the Riemannian differential of a curve of type `c` represented by function `f`
+at time `t` using the given backend.
+"""
+differential(::Any, ::Curve, ::Real, ::AbstractRiemannianDiffBackend)
+
+
+"""
+    gradient(f::Any, rf::RealField, p, backend::AbstractRiemannianDiffBackend = rgradient_backend())
+
+Compute the Riemannian gradient of a real field of type `rf` represented by function `f`
+at point `p` using the given backend.
+"""
+gradient(::Any, ::RealField, ::Any, ::AbstractRiemannianDiffBackend)
+
+function differential!(f::Any, c::Curve, X, t, backend::AbstractRiemannianDiffBackend)
+    return copyto!(X, differential(f, c, t, backend))
+end
+
+function gradient!(f, ft::RealField, X, p, backend::AbstractRiemannianDiffBackend)
+    return copyto!(X, gradient(f, ft, p, backend))
+end
+
+differential(f, c::Curve, p) = differential(f, c, p, rdifferential_backend())
+
+differential!(f, c::Curve, X, p) = differential!(f, c, X, p, rdifferential_backend())
+
+gradient(f, rf::RealField, p) = gradient(f, rf, p, rgradient_backend())
+
+gradient!(f, rf::RealField, X, p) = gradient!(f, rf, X, p, rgradient_backend())
+
+"""
+    RiemannianONBDiffBackend(
+        diff_backend::AbstractDiffBackend
+        retraction::AbstractRetractionMethod
+        inverse_retraction::AbstractInverseRetractionMethod
+        basis::Union{AbstractOrthonormalBasis,CachedBasis{<:AbstractOrthonormalBasis}},
+    ) <: AbstractRiemannianDiffBackend
+
+Riemannian differentiation based on differentiation in an [`AbstractOrthonormalBasis`](@ref)
+`basis` using specified `retraction`, `inverse_retraction` and using backend `diff_backend`.
+"""
+struct RiemannianONBDiffBackend{
+    TADBackend<:AbstractDiffBackend,
+    TRetr<:AbstractRetractionMethod,
+    TInvRetr<:AbstractInverseRetractionMethod,
+    TBasis<:Union{
```

Comment:
Why store both? The cached basis “knows” its internal ONB?

Reply:
What do you mean by both? It stores either a cached or a non-cached basis. It's a union to only let orthonormal cached bases.

Reply:
Ah, ok. Then I misread the code and its fine.